src/sds_data_model/_raster.py

```diff
@@ -0,0 +1,466 @@
+        expected_x_max (int): x maximum (width).
+        expected_y_max (int): y maxium (height).
+        nodata (Optional[float]): value that will fill the grid where
+            there is no data (if it is not `None`).
+
+    Returns:
+        DataArray: A `DataArray` the shape of the specified grid,
+        containing the data of the input `DataArray`.
+    """
+    _check_no_data(data_array, nodata)
+
+    offset = _get_origin_offset(data_array, expected_x_min, expected_y_max)
+
+    bng = full(
+        shape=(
+            array([expected_y_max, expected_x_max]) / data_array.rio.transform().a
+        ).astype("int"),
+        fill_value=data_array.rio.nodata,
+        dtype=data_array.dtype,
+    )
+
+    bng[
+        offset[0] : offset[0] + data_array.shape[0],
+        offset[1] : offset[1] + data_array.shape[1],
+    ] = data_array.data
+
+    return _create_data_array(
+        data_array=data_array,
+        data=bng,
+        geotransform=(
+            f"{expected_x_min} {data_array.rio.transform().a} 0 "
+            f"{expected_y_max} 0 {data_array.rio.transform().e}"
+        ),
+    )
+
+
+def _select_bands(
+    dataset: Dataset,
+    bands: Optional[Union[List[int], List[str]]],
+    engine: Optional[str],
+) -> List[DataArray]:
+    """Extracts each band from the input dataset.
+
+    Specific bands are selected using the `bands` arg if it is provided.
+    If the `engine` arg is rasterio, bands may have a `long_name` attr.
+    If this attr doesn't exist the position/index of the band is used as its name.
+
+    Args:
+        dataset (Dataset): Input dataset.
+        bands (Optional[Union[List[int], List[str]]]):
+            List of band numbers or names.
+        engine (Optional[str]): Engine used by open_dataset.
+
+    Returns:
+        List[DataArray]: List of `DataArray`s (bands).
+    """
+    if engine == "rasterio":
+        try:
+            band_names = dataset.band_data.long_name
+        except AttributeError:
+            band_names = range(1, len(dataset.band_data) + 1)
+        data_arrays = [
+            data_array.rename(f"{band_name}").drop("band")
+            for band_name, data_array in zip(band_names, dataset.band_data)
+        ]
+    else:
+        data_arrays = [dataset[array_name] for array_name in dataset.rio.vars]
+
+    if bands:
+        data_arrays = [
+            data_array
+            for data_array in data_arrays
+            if data_array.name in map(str, bands)
+        ]
+    return data_arrays
+
+
+def _resample_and_reshape(
+    dataset: Dataset,
+    expected_cell_size: int,
+    expected_x_min: int,
+    expected_x_max: int,
+    expected_y_max: int,
+    bands: Optional[Union[List[int], List[str]]],
+    categorical: Union[bool, Dict[Union[int, str], bool]],
+    nodata: Optional[float],
+    engine: Optional[str],
+) -> Dataset:
+    """Checks and conditionally changes the cell size and shape/extent \
+    of the input `Dataset` against the input values.
```

Comment:
As above, the first line of a function docstring should should be a one line (<88 character) description of the function.

```suggestion
    """Resamples and/or reshapes a `Dataset` with the wrong extent or cell size.
```